proposals/lip-0044.md

```diff
@@ -54,7 +54,6 @@ We define the following constants:
 | `SUBSTORE_PREFIX_VALIDATORS_KEYS` | bytes | 0x0000 | Substore prefix of the validators keys substore. |
 | `SUBSTORE_PREFIX_VALIDATOR_PARAMS` | bytes | 0x4000 | Substore prefix of the validator params substore. |
 | `SUBSTORE_PREFIX_BLS_KEYS` | bytes | 0x8000 | Substore prefix of the registered BLS keys substore. |
-| `SUBSTORE_PREFIX_GENESIS_DATA` | bytes | 0xc000 | Substore prefix of the genesis data substore. |
```

Comment:
Line 41: I don't understand what we mean that we check the existence of a BLS key in constant time. I guess it is logarithmic in the number of validators, instead of linear that it would be by checking all addresses if we did not have the registered BLS keys substore. Am I missing something?

Reply:
The basic assumption here is that state store reads are O(1) as it is basically a hash map (the actual rocks DB implementation uses caching and other advanced techniques so maybe this is a simplification). However, with this assumption checking if a BLS key is registered can be done in O(1) utilizing the "Registered BLS Keys Substore". This store contains redundant date in order to achieve this performance.

Reply:
Ok I think I see now. You assume that you use hashing therefore you can look up in expected time O(1). So it's just that you use the PublicKeyBLS as a key (instead of e.g., address) that allows you to achieve this? Otherwise by just scanning over all addresses to check for a specific BLS you would need \Omega(n) anyways. Is this the punchline?

Reply:
> So it's just that you use the PublicKeyBLS as a key (instead of e.g., address) that allows you to achieve this?
 
Yes. One does not need to scan through all addresses. Only one look up a given store key (the BLS key).